core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -249,12 +249,22 @@ public AtomicReference<Long> getWorkerId() {
   public UfsStatus[] listStatus(String path, ListOptions options) throws IOException {
     UfsStatus[] statuses = mListStatusCache.getIfPresent(path);
     if (statuses == null) {
-      // Not found in cache. Query the Under File System.
-      statuses = mUfs.listStatus(path, options);
-      // Add this into cache. Return value might be null if not found.
-      if (statuses != null) {
-        mListStatusCache.put(path, statuses);
-      }
+      // If not found, let's get it synchronously, avoiding multiple threads listing the same
+      // directory. If a directory contains a large number of files/dirs, the listStatus()
+      // needs a lot of memory for the results.
+      synchronized (mListStatusCache) {
```

Comment:
not quite sure if we lock at directory level, otherwise all list status requests of different directories block each other

Reply:
Yes. After I wrote this, I had the same question.
We need finer grained lock/synchronization semantics.
For example, we only lock a particular dir at a given time.

Reply:
Make sense!